sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala

```diff
@@ -0,0 +1,258 @@
+
+  override def dataType: DataType = ArrayType(MapType(ArrayType(StringType), DoubleType))
+
+  override def prettyName: String = "ngrams"
+
+  override def serialize(obj: NGramBuffer): Array[Byte] = {
+    NGrams.serializer.serialize(obj)
+  }
+
+  override def deserialize(bytes: Array[Byte]): NGramBuffer = {
+    NGrams.serializer.deserialize(bytes)
+  }
+}
+
+object NGrams {
+
+  val DEFAULT_ACCURACY: Int = 1000
+
+  private def getAccuracy(kExpression: Expression, accuracyExpression: Expression): Expression = {
+    val accuracy = accuracyExpression.eval().asInstanceOf[Int]
+    val k = kExpression.eval().asInstanceOf[Int]
+    Literal(accuracy.max(DEFAULT_ACCURACY / k))
+  }
+
+  val kryoSerializer: KryoSerializer = new KryoSerializer(new SparkConf())
+
+  import collection.JavaConverters._
+
+  class NGramBuffer(val n: Int,
+                    val k: Int,
+                    val precisionFactor: Int,
+                    val frequencyMap: HashMap[Vector[UTF8String], Double]) {
+    def add(ng: Vector[UTF8String]): Unit = {
+      var currentFrequency: Double = frequencyMap.get(ng)
+      if (currentFrequency == null.asInstanceOf[Double]) {
+        currentFrequency = 1.0D
+      } else {
+        currentFrequency += 1
+      }
+      frequencyMap.put(ng, currentFrequency)
+    }
+
+    def merge(other: NGramBuffer): Unit = {
+      other.frequencyMap.asScala.foreach((keyValuePair: (Vector[UTF8String], Double)) => {
+        val key = keyValuePair._1
+        val value = keyValuePair._2
+        val originalValue = frequencyMap.getOrDefault(key, 0.0D)
+        frequencyMap.put(key, originalValue + value)
+      })
+    }
+
+    def sortWithTwoFields(frequencyDescend: Boolean)
+                         (keyWithFrequency: Tuple2[Vector[UTF8String], Double],
+                          keyWithFrequency2: Tuple2[Vector[UTF8String], Double]): Boolean = {
+      if (keyWithFrequency._2 != keyWithFrequency2._2) {
+      (keyWithFrequency._2 < keyWithFrequency2._2) ^ frequencyDescend
+      }
+      else {
+        val keyVector = keyWithFrequency._1
+        val keyVector2 = keyWithFrequency2._1
+        for (i <- 0 until keyVector.length) {
+          if (keyVector(i) != keyVector2(i)) {
+            return (keyVector(i).compare(keyVector2(i))) < 0
+          }
+        }
+        true
+      }
+    }
+
+    def trim(): Unit = {
+      if (frequencyMap.size() > 2 * k * precisionFactor) {
+        val orderedWithIndex = frequencyMap.asScala.iterator.toVector.
+          sortWith(sortWithTwoFields(frequencyDescend = false)).zipWithIndex
+        orderedWithIndex.takeWhile(_._2 < frequencyMap.size() - k * precisionFactor).map(_._1).
+          foreach(keyValuePair => frequencyMap.remove(keyValuePair._1))
+      }
+    }
+
+    def getTopKNGrams(): Seq[(Vector[UTF8String], Double)] = {
+      frequencyMap.asScala.iterator.toVector.sortWith(sortWithTwoFields(frequencyDescend = true)).
+        zipWithIndex.takeWhile(_._2 < k).map(_._1)
```

Comment:
You can simply do:
 
    frequencyMap.asScala.iterator.toVector.sortWith(sortWithTwoFields(frequencyDescend = true)).take(k)